tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas3_overwriting_gemm_kk.hpp

```diff
@@ -0,0 +1,271 @@
+         class ElementType_BT,
+         std::experimental::extents<>::size_type numRows_BT,
+         std::experimental::extents<>::size_type numCols_BT,
+         class NestedLayout_B,
+	 class Accessor_BT,
+         class ElementType_C,
+         std::experimental::extents<>::size_type numRows_C,
+         std::experimental::extents<>::size_type numCols_C,
+         class Layout_C,
+	 class Accessor_C,
+	 /* requires */
+	 (Layout_A::template mapping<std::experimental::extents<numRows_A, numCols_A>>::is_always_unique() &&
+	  NestedLayout_B::template mapping<std::experimental::extents<numRows_BT, numCols_BT>>::is_always_unique() &&
+	  Layout_C::template mapping<std::experimental::extents<numRows_C, numCols_C>>::is_always_unique()
+	  ))
+void matrix_product(
+  kokkos_exec<ExeSpace> /*kexe*/,
+  std::experimental::mdspan<
+    ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A, Accessor_A
+  > A,
+  std::experimental::mdspan<
+    ElementType_BT, std::experimental::extents<numRows_BT, numCols_BT>,
+    std::experimental::linalg::layout_transpose<NestedLayout_B>, Accessor_BT
+  > BT,
+  std::experimental::mdspan<
+    ElementType_C, std::experimental::extents<numRows_C, numCols_C>, Layout_C, Accessor_C
+  > C)
+{
+
+  // note that BT = B^T so BT carries the tranpose effect
+
+  // preconditions
+  if ( A.extent(1) != BT.extent(0) ){
+    throw std::runtime_error("KokkosBlas: gemm_C_ABT_product: A.extent(1) != BT.extent(0) ");
+  }
+  if ( A.extent(0) != C.extent(0) ){
+    throw std::runtime_error("KokkosBlas: gemm_C_ABT_product: A.extent(0) != C.extent(0) ");
+  }
+  if ( BT.extent(1) != C.extent(1) ){
+    throw std::runtime_error("KokkosBlas: gemm_C_ABT_product: BT.extent(1) != C.extent(1) ");
+  }
+
+  // mandates
+  ov_gemm_impl::static_extent_match(A.static_extent(1),  BT.static_extent(0));
+  ov_gemm_impl::static_extent_match(A.static_extent(0),  C.static_extent(0));
+  ov_gemm_impl::static_extent_match(BT.static_extent(1), C.static_extent(1));
+
+  // note that the conversion to view does NOT carry the transpose
+  auto A_view = Impl::mdspan_to_view(A);
+  auto B_view = Impl::mdspan_to_view(BT);
+  auto C_view = Impl::mdspan_to_view(C);
+
+  namespace stdexp = std::experimental;
+
+  constexpr bool doing_C_equal_A_BT =
+    std::is_same_v<Accessor_A, stdexp::default_accessor<ElementType_A>> &&
+    std::is_same_v<Accessor_BT, stdexp::default_accessor<ElementType_BT>> &&
+    std::is_same_v<Accessor_C, stdexp::default_accessor<ElementType_C>>;
+
+  if constexpr(doing_C_equal_A_BT){
+    Impl::signal_kokkos_impl_called("gemm_C_ABT_product");
+
+    const auto alpha = static_cast<typename decltype(A_view)::value_type>(1);
+    const auto beta  = static_cast<typename decltype(C_view)::value_type>(0);
+    // we need to explicitly pass "T" because B_view does NOT carry the transpose
+    KokkosBlas::gemm("N", "T", alpha, A_view, B_view, beta, C_view);
+  }
+  // todo: handle cases stemming from non-trivial accessors
```

Comment:
I have some ideas for this -- I'll file an issue.

Reply:
my idea was to progressively handle all those cases one at a time. but maybe it is worth it to figure out a better design to handle all the subcases.